### PR TITLE
Question answering

### PR DESCRIPTION
**Pull request by delphieritas**
Question_Answering.train() will return a table containing the Q and A. 
Question_Answering.evaluate() is not applicable, thus return a pseudo float(1).
Might need to overwrite Question_Answering.predict() for user defined prediction.

### CODE REVIEW

File: examples/models/question_answering/Question_Answering.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+        confidence_scores.append(float(confidence))
+
+    maxIdx = np.argmax(confidence_scores)
+    confidence = confidence_scores[maxIdx]
+    best_answer = answers[maxIdx]
+    best_paragraph = paper[maxIdx]
+    print (best_answer, confidence, best_paragraph)
+    return best_answer, confidence, best_paragraph
+
+class Question_Answering(BaseModel):
```
Comment (nudles) [inline examples/models/question_answering/Question_Answering.py]: The name should be either XXXYYYZZZ or xxx_yyy_zzz. There is no XXX_YYY

File: examples/models/question_answering/Question_Answering.py [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+        qa_tokenizer = AutoTokenizer.from_pretrained(BERT_SQUAD)
+        qa_model = AutoModelForQuestionAnswering.from_pretrained(BERT_SQUAD)
+
+        qa_model = qa_model.to(torch_device)
+        qa_model.eval()
+        return qa_model
+
+
+    def train(self, dataset_path=None, shared_params: Optional[Params] = None, **train_args):
+        self._model = self._create_model(scratch = False)
```
Comment (nudles) [inline examples/models/question_answering/Question_Answering.py]: this model does not need any training.
we just do prediction.
hence, the train() should be empty.

File: examples/models/question_answering/Question_Answering.py [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+        num_papers = 5
+
+        data = {"Task1":
```
Comment (nudles) [inline examples/models/question_answering/Question_Answering.py]: the data provided by kaggle notebook is just an example. they should not appear in the python code.
you can put them into a readme.md file to show the users what kind of questions can be asked.
users will use the front-end to send the questions to the QAS model.

File: examples/models/question_answering/Question_Answering.py [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+            # save as output
+            results_table.to_csv("test/task_{}_results.tsv".format(task_cnt))
+            task_cnt += 1
+        print(results_table, task_cnt)
+        return results_table
+
+    def evaluate(self, dataset_path):
+        return float(1)
+
+    def predict(self, queries=None):
```
Comment (nudles) [inline examples/models/question_answering/Question_Answering.py]: the predict() function receives the user provided question and returns the answer.
refer to this pic I shared before.
WeChat Image_20200517110217

**Comment by chrishkchris** [comment]
FYI, In this PR, the linting errors in other files has already been fixed in master branch
Therefore, the only linting error in this PR is 
```
************* Module Question_Answering
examples/models/question_answering/Question_Answering.py:465:22: E1120: No value for argument 'scratch' in method call (no-value-for-parameter)
```

Starting from next PR, should use git pull to fetch the latest update from nusdbsystem:master first, in order to remove the linting errors that are already fixed

**Comment by chrishkchris** [comment]
What else do you need to do (e.g. TO DO list) in this PR before it is ready for use?

Thao has finished the web app for Covid-19, so you can test the deployed web app after the training https://github.com/nusdbsystem/singa-easy/pull/6

**Comment by nudles** [comment]
@delphieritas have you address the issues? can you push your new commits?
Have you tested your code with Thao's PR for the front-end.

**Comment by nudles** [comment]
@delphieritas is it done? can you push your code?

**Comment by delphieritas** [comment]
@nudles 
yep. the model is done with pre-specified issues fixed. will PR now

File: examples/models/question_answering/Question_Answering.py [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+                            "journal_score": journal_score,
+                            "paper_citations_count": paper_citations_count,
+                        }
+                        answers_list.append(paper_answer)
+                dict_queries[query] = answers_list
+            dict_task_quest['queries'] = dict_queries
+            dict_task_quest['task_name'] = data[task]['area']
+            final_dict[task] = dict_task_quest
+        full_code = final_dict
+        # return getHtmlCode(full_code, layoutStyle())
```
Comment (nudles) [inline examples/models/question_answering/Question_Answering.py]: why not call getHtmlCode()?
Can the front-end parse and display the returned table?

Comment (delphieritas) [reply examples/models/question_answering/Question_Answering.py]: Now it supports getHtmlCode(). Just by uncommenting this line while commenting out the following lines.
I optimised admin.py and dev.py files for singa-auto, to better run model efficiently & create_dataset() regarding to 'task == "question_answering", and made PR #65 .

**Pull request merged**